EIP-7594: Decouple network subnets from das-core (#3832)
ethereum/consensus-specs: specs/_features/eip7594/das-core.md

| - | - | - |
| `RowIndex` | `uint64` | Row identifier in the matrix of cells |
| `ColumnIndex` | `uint64` | Column identifier in the matrix of cells |

> | `CustodyIndex` | `uint64` | Custody group identifier in the matrix of cells |

Reviewer: This is a bit wrong in my opinion. It's an index into a set of groups & doesn't involve the matrix.
```suggestion
| `CustodyIndex` | `uint64` | Custody group identifier in the set of custody groups |
```

| `NUMBER_OF_COLUMNS` | `uint64(CELLS_PER_EXT_BLOB)` (= 128) | Number of columns in the extended data matrix |
| `MAX_CELLS_IN_EXTENDED_MATRIX` | `uint64(MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS)` (= 768) | The data size of `ExtendedMatrix` |

### Custody setting

| Name | Value | Description |
| - | - | - |
| `SAMPLES_PER_SLOT` | `8` | Number of `DataColumnSidecar` random samples a node queries per slot |

> | `NUMBER_OF_CUSTODY_GROUPS` | `128` | Number of column groups available for nodes to custody |

Reviewer: Yes, a custody group is a collection of columns, but calling it a "column group" is a inconsistent.
```suggestion
| `NUMBER_OF_CUSTODY_GROUPS` | `128` | Number of custody groups available for nodes to custody |
```

Author: Yeah, maybe I missed that during the rebase

| `NUMBER_OF_COLUMNS` | `uint64(CELLS_PER_EXT_BLOB)` (= 128) | Number of columns in the extended data matrix. |
| `MAX_CELLS_IN_EXTENDED_MATRIX` | `uint64(MAX_BLOBS_PER_BLOCK * NUMBER_OF_COLUMNS)` (= 768) | The data size of `ExtendedMatrix`. |

### Custody setting

| Name | Value | Description |
| - | - | - |
| `SAMPLES_PER_SLOT` | `8` | Number of `DataColumnSidecar` random samples a node queries per slot |

> | `NUMBER_OF_CUSTODY_GROUPS` | `32` | Number of column groups nodes will custody. |

Author: I think you meant this one?
```suggestion
| `NUMBER_OF_CUSTODY_GROUPS` | `32` | Number of column groups available for nodes to custody. |
```

> | `NUMBER_OF_CUSTODY_GROUPS` | `32` | Number of column groups nodes will custody. |

Reviewer: Oops, yes that's that I meant. Thanks!

### Helper functions

#### `get_custody_groups`

```python
def get_custody_groups(node_id: NodeID, custody_group_count: uint64) -> Sequence[CustodyIndex]:
    assert custody_group_count <= NUMBER_OF_CUSTODY_GROUPS

    custody_groups: List[uint64] = []
    current_id = uint256(node_id)
    while len(custody_groups) < custody_group_count:
        custody_group = CustodyIndex(
            bytes_to_uint64(hash(uint_to_bytes(uint256(current_id)))[0:8])
            % NUMBER_OF_CUSTODY_GROUPS
        )
        if custody_group not in custody_groups:
            custody_groups.append(custody_group)
        if current_id == UINT256_MAX:
            # Overflow prevention
            current_id = NodeID(0)
        current_id += 1

    assert len(custody_groups) == len(set(custody_groups))
    return sorted(custody_groups)
```

#### `compute_columns_for_custody_group`

```python
def compute_columns_for_custody_group(custody_group: CustodyIndex) -> Sequence[ColumnIndex]:

>     columns_per_group = NUMBER_OF_COLUMNS // NUMBER_OF_CUSTODY_GROUPS

Author: That makes sense. Thanks.

### Custody requirement

> Columns are grouped in custody groups. Nodes custodying a custody group are supposed to custody all the columns in that group.

Author: How about?
```suggestion
Columns are grouped into custody groups. Nodes custodying a custody group MUST custody all the columns in that group.
```

> Columns are grouped in custody groups. Nodes custodying a custody group are supposed to custody all the columns in that group.

Reviewer: Yup that sounds good too 👍

However, for simplicity, we don't assign row custody assignments to nodes in the current design.

### Why don't we rotate custody over time?

To start with a simple, stable backbone, for now, we don't shuffle the custody assignments via the deterministic custody selection helper `get_custody_groups`. However, staggered rotation likely needs to happen on the order of the pruning period to ensure subnets can be utilized for recovery. For example, introducing an `epoch` argument allows the function to maintain stability over many epochs.

### Does having a lot of column subnets make the network unstable?

> No, the number of subnets doesn't really matter. What matters to the network stability is the number of nodes and the churn rate in the network. If the number of the nodes is too low, it's likely to have a network partition when some nodes are down. For the churn rate, if the churn rate is high, we even need to have a higher number of nodes, since nodes are likely to be turned off more often.

Reviewer 2: Having a higher subnet count may require us to bump up our target peer count, as there would be a higher probability that nodes have 0 peers in a given column subnet, and will have to do more frequent discovery.

@dknopik  has done some interesting [simulation on PeerDAS](https://hackmd.io/@dknopik/epf-week16) with 128 column subnets, and the LH-only network (without any supernodes) performs poorly with our current peer count (100), but much better when when target peers is increased to 300. With a small peer count, there's a higher chance that the node doesn't have peers across all column subnets, and fails to publish columns to the network. However, increasing peer count to 300 does increase the resource consumption in LH significantly, and our preference is to increase our peer gradually to ~120. I think there's some work we need to do to improve our peer selection logic for PeerDAS but thought it might be worth mentioning.

I'm wondering if we need to choose a column subnet count where configuring `CUSTODY_REQUIREMENT` to a lower number won't cause the network to fall apart, with the existing target peer count?

With this, I'm convinced that decoupling network and das-core do seems like a useful thing - because without decoupling, changing any of the parameters would have impact to _BOTH_ security and networking. If we decouple networking and das-core, we could determine an optimal networking parameters and test different custody parameters, and vice versa.

Author: > Having a higher subnet count may require us to bump up our target peer count, as there would be a higher probability that nodes have 0 peers in a given column subnet, and will have to do more frequent discovery.

According to this PR, the target peer count depends on the number of custody groups, right? Yes, before this PR it depends on the number of subnets, but this PR changes the structure and the terminology. WDYT?

However, for simplicity, we don't assign row custody assignments to nodes in the current design.

### Why don't we rotate custody over time?

To start with a simple, stable backbone, for now, we don't shuffle the custody assignments via the deterministic custody selection helper `get_custody_groups`. However, staggered rotation likely needs to happen on the order of the pruning period to ensure subnets can be utilized for recovery. For example, introducing an `epoch` argument allows the function to maintain stability over many epochs.

### Does having a lot of column subnets make the network unstable?

> No, the number of subnets don't really matter. What matters to the network stability is the number of nodes and the churn rate in the network.

Reviewer: ```suggestion
No, the number of subnets doesn't really matter. What matters to network stability is the number of nodes and the churn rate in the network.
```

Reviewer 3: If we're going to have this paragraph at all I think it would be good to expand on it a bit, for example on the "what matters" part

Reviewer: Agreed. I also thought it was a bit light on details.
